src/App.tsx

```diff
@@ -1,24 +1,240 @@
+      .then(setTodos)
+      .catch(() => setErrorMessage(ErrorMessage.LoadTodos));
+  }, []);
+
+  useEffect(() => {
+    const timerId = window.setTimeout(() => {
+      clearErrorMessage();
+    }, 3000);
+
+    return () => window.clearTimeout(timerId);
+  }, [errorMessage]);
+
+  const startProcessingTodo = (todo: Todo) => {
+    setProcessingIds(currentIds => [...currentIds, todo.id]);
+  };
+
+  const stopProcessingTodo = (todo: Todo) => {
+    setProcessingIds(currentIds => currentIds.filter(id => id !== todo.id));
+  };
+
+  const addTodoToState = (newTodo: Todo) => {
+    setTodos(currentTodos => [...currentTodos, newTodo]);
+  };
+
+  const replaceTodoInState = (updatedTodo: Todo) => {
+    setTodos(currentTodos => {
+      return currentTodos.map(currentTodo => (
+        updatedTodo.id === currentTodo.id
+          ? updatedTodo
+          : currentTodo
+      ));
+    });
+  };
+
+  const removeTodoFromState = (todo: Todo) => {
+    setTodos(currentTodos => {
+      return currentTodos.filter(({ id }) => id !== todo.id);
+    });
+  };
+
+  const handleAddTodo = useCallback((title: string) => {
+    clearErrorMessage();
+    setTempTodo({
+      id: 0,
+      userId: 0,
+      title,
+      completed: false,
+    });
+
+    return todoService.createTodo(title)
+      .then(addTodoToState)
+      .catch(() => {
+        setErrorMessage(ErrorMessage.AddTodo);
+
+        throw new Error(ErrorMessage.AddTodo);
+      })
+      .finally(() => setTempTodo(null));
+  }, []);
+
+  const handleDeleteTodo = useCallback((todo: Todo) => {
+    clearErrorMessage();
+    startProcessingTodo(todo);
+
+    todoService.deleteTodo(todo.id)
+      .then(() => removeTodoFromState(todo))
+      .catch(() => setErrorMessage(ErrorMessage.DeleteTodo))
+      .finally(() => stopProcessingTodo(todo));
+  }, []);
+
+  const handleClearCompletedTodos = useCallback(() => {
+    todos
+      .filter(({ completed }) => completed)
+      .forEach(handleDeleteTodo);
+  }, [todos]);
+
+  const handleToggleTodo = useCallback((todo: Todo) => {
+    clearErrorMessage();
+    startProcessingTodo(todo);
+
+    todoService.toggleTodo(todo.id, todo.completed)
+      .then(replaceTodoInState)
+      .catch(() => setErrorMessage(ErrorMessage.UpdateTodo))
+      .finally(() => stopProcessingTodo(todo));
+  }, []);
+
+  const handleToggleAllTodos = useCallback(() => {
+    const hasActive = todos.some(({ completed }) => !completed);
+
+    todos.forEach(todo => {
+      if (hasActive && todo.completed) {
+        return;
+      }
+
+      handleToggleTodo(todo);
+    });
+  }, [todos]);
+
+  const handleEditTodo = useCallback((todo: Todo, title: string) => {
+    clearErrorMessage();
+    startProcessingTodo(todo);
+
+    if (!title) {
+      return todoService.deleteTodo(todo.id)
+        .then(() => removeTodoFromState(todo))
+        .catch(() => {
+          setErrorMessage(ErrorMessage.DeleteTodo);
+
+          throw new Error(ErrorMessage.DeleteTodo);
```

Comment:
Why you need to throw new error here?

Reply:
To catch it in child component and do something there too.